src/device-manager/WeaveDataManagementClient.cpp

```diff
@@ -985,7 +985,7 @@ void WdmClient::ClientEventCallback(void * const aAppState, SubscriptionClient::
         WeaveLogDetail(DataManagement, "Client->kEvent_OnSubscriptionTerminated. Reason: %u, peer = 0x%" PRIX64 "\n",
                        aInParam.mSubscriptionTerminated.mReason, aInParam.mSubscriptionTerminated.mClient->GetPeerNodeId());
         pWdmClient->mpSubscriptionClient->AbortSubscription();
-        err = WEAVE_ERROR_INCORRECT_STATE;
+        err = aInParam.mSubscriptionTerminated.mReason;
```

Comment:
This isn't going to be very useful for applications in the case where the subscription terminated due to an error from the publisher.  In that case, the information returned by the publisher is in the status report, and mSubscriptionTerminated.mReason will be WEAVE_ERROR_STATUS_REPORT_RECEIVED.  There needs to be a way to return the status report information back to the application.

Reply:
> This isn't going to be very useful for applications in the case where the subscription terminated due to an error from the publisher. In that case, the information returned by the publisher is in the status report, and mSubscriptionTerminated.mReason will be WEAVE_ERROR_STATUS_REPORT_RECEIVED. There needs to be a way to return the status report information back to the application.

Yep, agree, mSubscriptionTerminated has already included profile id and status code, I think we can return 4 variable(error code, isStatusCodeValid, statusProfileId, statusCode) to application, and we can add new refresh callback, OnRefreshError

        struct
        {
            WEAVE_ERROR mReason;

            bool mIsStatusCodeValid;
            uint32_t mStatusProfileId;
            uint16_t mStatusCode;
            ReferencedTLVData * mAdditionalInfoPtr;
            bool mWillRetry;

            SubscriptionClient * mClient;
        } mSubscriptionTerminated;

Reply:
I would prefer if we can follow the conventions used by WeaveDeviceManager. On the Java side, we normally get an onError callback with a Throwable as the parameter type. The concrete type of that Throwable will be either WeaveDeviceException (for profile/status errors) or WeaveDeviceManagerException (for status code errors).

Reply:
@yunhanw-google Kevin's approach is right.  In C++ pass the reason, status profile and status code back via the onError callback.  Then in the Java wrapper code, translate these into WeaveDeviceException or WeaveDeviceManagerException.

No need to pass isStatusCodeValid to C++ callback.  If isStatusCodeValid == true, return profile and status code and set err = WEAVE_ERROR_STATUS_REPORT_RECEIVED.  If false, just return err = mReason and set profile and code to 0.

In Java wrapper, use err == WEAVE_ERROR_STATUS_REPORT_RECEIVED to decide which type of exception to return.